File: lib/services/geocoding_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class AddressDetails {
  final String? road;
  final String? suburb;
  final String? city;
  final String? state;
  final String? country;
  final String? displayName;

  const AddressDetails({
    this.road,
    this.suburb,
    this.city,
    this.state,
    this.country,
    this.displayName,
  });

  factory AddressDetails.fromJson(Map<String, dynamic> json, String displayName) {
    final addr = json['address'] as Map<String, dynamic>? ?? {};
    return AddressDetails(
      road: addr['road'] as String? ?? addr['pedestrian'] as String?,
      suburb: addr['suburb'] as String? ?? addr['neighbourhood'] as String?,
      city: addr['city'] as String? ?? addr['town'] as String? ?? addr['village'] as String?,
      state: addr['state'] as String?,
      country: addr['country'] as String?,
      displayName: displayName,
    );
  }
}

class GeocodingSearchResult {
  final double lat;
  final double lng;
  final AddressDetails address;

  const GeocodingSearchResult({
    required this.lat,
    required this.lng,
    required this.address,
  });
}

class GeocodingService {
  static const String userAgent = 'CoolNet-AI-Flutter/1.0';

  static Future<AddressDetails?> reverseGeocode(double lat, double lng) async {
    try {
      final Uri uri = Uri.parse(
          'https://nominatim.openstreetmap.org/reverse?lat=$lat&lon=$lng&format=json&addressdetails=1');
      final response = await http.get(uri, headers: {'User-Agent': userAgent});
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        return AddressDetails.fromJson(data, data['display_name'] as String? ?? '');
      }
    } catch (e) {
      // Graceful fallback
    }
    return null;
  }

  static Future<List<GeocodingSearchResult>> searchLocation(String query) async {
    if (query.trim().length < 3) return [];
    try {
      final Uri uri = Uri.parse(
          'https://nominatim.openstreetmap.org/search?q=${Uri.encodeComponent(query)}&format=json&limit=5&addressdetails=1');
      final response = await http.get(uri, headers: {'User-Agent': userAgent});
      if (response.statusCode == 200) {
        final List list = jsonDecode(response.body) as List;
        return list.map((item) {
          final map = item as Map<String, dynamic>;
          return GeocodingSearchResult(
            lat: double.parse(map['lat'].toString()),
            lng: double.parse(map['lon'].toString()),
            address: AddressDetails.fromJson(map, map['display_name'] as String? ?? ''),
          );
        }).toList();
      }
    } catch (e) {
      // Graceful fallback
    }
    return [];
  }
}

File: src/lib/services/geocoding-service.ts
import { AddressDetails } from "@/lib/types";

const USER_AGENT = "CoolNet-AI-Climate-Dashboard/1.0";

/**
 * Perform reverse geocoding to retrieve structured address from coordinates
 */
export async function reverseGeocode(lat: number, lng: number): Promise<AddressDetails | null> {
  try {
    const url = `https://nominatim.openstreetmap.org/reverse?lat=${lat}&lon=${lng}&format=json&addressdetails=1`;
    const response = await fetch(url, {
      headers: {
        "User-Agent": USER_AGENT
      }
    });

    if (!response.ok) {
      throw new Error(`Reverse geocode returned status ${response.status}`);
    }
    const data = await response.json();
    if (!data || !data.address) return null;

    const address = data.address;
    return {
      road: address.road || address.pedestrian,
      neighbourhood: address.neighbourhood,
      suburb: address.suburb,
      city_district: address.city_district || address.subdistrict,
      city: address.city,
      town: address.town,
      village: address.village,
      county: address.county,
      state: address.state,
      postcode: address.postcode,
      country: address.country,
      country_code: address.country_code,
      displayName: data.display_name
    };
  } catch (error) {
    console.error("Geocoding: Reverse geocode failed", error);
    return null;
  }
}

/**
 * Searches Nominatim for matching names and addresses
 */
export async function searchLocation(query: string): Promise<{ lat: number; lng: number; address: AddressDetails }[]> {
  if (!query || query.trim().length < 3) return [];
  try {
    const url = `https://nominatim.openstreetmap.org/search?q=${encodeURIComponent(query)}&format=json&limit=5&addressdetails=1`;
    const response = await fetch(url, {
      headers: {
        "User-Agent": USER_AGENT
      }
    });

    if (!response.ok) {
      throw new Error(`Location search returned status ${response.status}`);
    }
    const data = await response.json();
    if (!Array.isArray(data)) return [];

    return data.map((item: any) => {
      const address = item.address || {};
      return {
        lat: parseFloat(item.lat),
        lng: parseFloat(item.lon),
        address: {
          road: address.road || address.pedestrian,
          neighbourhood: address.neighbourhood,
          suburb: address.suburb,
          city_district: address.city_district || address.subdistrict,
          city: address.city,
          town: address.town,
          village: address.village,
          county: address.county,
          state: address.state,
          postcode: address.postcode,
          country: address.country,
          country_code: address.country_code,
          displayName: item.display_name
        }
      };
    });
  } catch (error) {
    console.error("Geocoding: Location search failed", error);
    return [];
  }
}
